Output 32-bit float samples to preserve dynamic range and prevent clipping

Falcosoft opened the pull request:
Added '-af' option to output 32-bit float samples in order to preserve dynamic range and prevent clipping at the same time.

Falcosoft committed:
Output 32-bit float samples to preserve dynamic range and prevent clipping

Added '-af' option to output 32-bit float samples in order to preserve dynamic range and prevent clipping at the same time.

giulioz commented on src/mcu.cpp:
```diff
@@ -1184,22 +1207,35 @@ void MCU_CloseAudio(void)
+    else if (sample_buffer_float)
+    {
+        const float divRec = 1 / 32768.0f;
+        sample[0] >>= 14;
```
There should be no need for bit shift if you divide for `2147483648.0` instead, and you could also keep more precision

Falcosoft replied:
HI,
If you inspect sample[0] and sample[1] at that place you will notice that if you right shift with 13 you will get samples that has a granulate of 2 if you right shift with 12 you will get  agranulate of 4 and so on. So currently with right shift 14 you get the maximum resolution. So you will not get more resolution with fewer right shifts. 
Moreover the float reolution is not infinite. By dividing with 2^31 you will get less precise results than with exact right shifts and then dividing with 2^15 since 2^15 and its multiplicants can be exactly represented in 32-bit float.

giulioz replied:
makes sense, thanks for the explanation!

Falcosoft replied:
Hi,
Actually you were right in the sense that the 2 shifts can be skipped and the result is the same. So no more precision, but somewhat better performance. It's because you do not have to divide by 2147483648.0 but only 536870912.0.

So overall this produces the same numbers as the current code:
```
void MCU_PostSample(int *sample)
{
    const float divRec = 1 / 536870912.0f;
    sample_buffer[sample_write_ptr + 0] = sample[0] * divRec;
    sample_buffer[sample_write_ptr + 1] = sample[1] * divRec;     

    sample_write_ptr = (sample_write_ptr + 2) % audio_buffer_size;
} 
```

Falcosoft commented on src/mcu.cpp:
```diff
@@ -1350,7 +1396,13 @@ int main(int argc, char *argv[])
+        }
+
+        if (audioFormat == AudioFormat::FLOAT32 && !isBufferSet)
+        {
```
In that case the argument order would be mandatory (that is -af has to be set after -ab) otherwise at 
(!strcmp(argv[i], "-af")) you cannot know if -ab is not set at all or only will be set after -af.
The if (audioFormat == AudioFormat::FLOAT32 && !isBufferSet) check is added to make sure that only change the dafault if -ab is not set manually.

Falcosoft committed:
Optimized process/thread priority for Windows

It significantly improves the consistency of audio output (less pauses, crackles) when the emulator is used together with other programs.  Tested with Win10 x64, Win7 x64, WinXP x86.

Falcosoft committed:
simpler int to float conversion

Falcosoft committed:
separated audio callbacks for short/float types

Wohlstand commented on src/mcu.cpp:
```diff
@@ -924,6 +930,11 @@ void MCU_WorkThread_Unlock(void)
 {
+
+#ifdef _WIN32
+    SetThreadPriority(GetCurrentThread(), THREAD_PRIORITY_HIGHEST);
```
Be careful! You can set the high priority, but not so much, on really slow machines this will make a hard stuck.

Falcosoft replied:
> Be careful! You can set the high priority, but not so much, on really slow machines this will make a hard stuck.

Hi,
THREAD_PRIORITY_HIGHEST only sets priority 2 points above the priority class. In case of default process priority class (8) means a priority of 10 that can never cause a hard stuck on any PC that is capable to run Nuked-SC55.

THREAD_PRIORITY_TIME_CRITICAL can cause such a problem since it sets priority of 15 for all normal priority calesses and priority of 31 for HIGH_PRIORITY_CLASS!
